Check the user's ansible.cfg for role/collection paths.

wenottingham opened the pull request:
There's no other way to add our new paths reliably without breaking things.

wenottingham committed:
Check the user's ansible.cfg for role/collection paths.

There's no other way to add our new paths reliably without breaking things.

wenottingham commented:
for https://github.com/ansible/awx/issues/4675. Lightly tested by editing roles_path in /etc/ansible/ansible.cfg and seeing if it shows up.

AlanCoding commented on awx/main/utils/ansible.py:
```diff
@@ -97,3 +98,14 @@ def could_be_inventory(project_path, dir_path, filename):
+def read_ansible_config(project_path, variables_of_interest):
+    fnames = [ os.path.join(project_path, 'ansible.cfg'), '/etc/ansible/ansible.cfg']
+    parser = ConfigParser()
+    conf = parser.read(fnames)
```
oh, didn't realize this could take a list, this should work just fine.

AlanCoding commented on awx/main/utils/ansible.py:
```diff
@@ -97,3 +98,14 @@ def could_be_inventory(project_path, dir_path, filename):
+    conf = parser.read(fnames)
+    values = {}
+    for var in variables_of_interest:
+        if var in conf['default']:
```
will need to look into if it's possible to throw a KeyError or something like that

wenottingham replied:
from the config? We wouldn't want to throw the keyerror ourselves.

AlanCoding replied:
```
result_traceback:
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 1209, in run
    private_data_files=private_data_files)
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/tasks.py", line 1481, in build_env
    config_values = read_ansible_config(job.project.get_project_path(), list(map(lambda x: x[1], path_vars)))
  File "/var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/main/utils/ansible.py", line 109, in read_ansible_config
    if var in conf['default']:
TypeError: list indices must be integers or slices, not str
```

don't know why

will work on next week

wenottingham replied:
s/default/defaults/, fixed.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -1473,14 +1474,22 @@ def build_env(self, job, private_data_dir, isolated=False, private_data_files=No
+            elif config_setting in config_values:
+                for path in config_values[config_setting].split(':'):
+                    if path not in paths:
+                        paths = [config_values[config_setting]] + paths
```
I'm thinking this might reverse the order of what was in `config_values[config_setting]`, will need to dig into that, if it is doing that, fix will be simple

AlanCoding replied:
Could you do a simple test using `role_paths=path1,path2` (either in the global location or project location) and see that the resultant `job_env` has the right ordering? Have you confirmed this manually?

AlanCoding replied:
oh wait, I see it now, you're not adding `path`, but adding `config_values[config_setting]` without the split.

This still isn't really airtight logic, I'd expect it to add duplicates in that case. Because you can't reliably check `if path not in paths` because you add entries to `paths` where the single entry contains multiple paths. While this will initially _work_, it will duplicate the entries.

AlanCoding commented:
Aside from the extremely minor comment I left about ordering, I see no reason we couldn't move ahead with this. @ryanpetrello I don't think you're going to like this, but I think we gotta do it.

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -1473,14 +1474,22 @@ def build_env(self, job, private_data_dir, isolated=False, private_data_files=No
+            ('ANSIBLE_COLLECTIONS_PATHS', 'collections_path', 'requirements_collections', '~/.ansible/collections:/usr/share/ansible/collections'),
+            ('ANSIBLE_ROLES_PATH', 'roles_path', 'requirements_roles', '~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles'))
+
+        config_values = read_ansible_config(job.project.get_project_path(), list(map(lambda x: x[1], path_vars)))
```
At this point, we've already set up the temporary directory and all the stuff in it. So if people wanted to avoid parsing the config as much as Earthly possible, we could avoid it in the event that the collection & role directories were empty.

matburt replied:
My only concern here would be issues parsing the file causing problems with the task executor causing an ambigious error. We might want to catch any exception that comes out of this and raise it as the `reason` a job failed.

wenottingham replied:
updated

wenottingham replied:
I don't think we want to `raise` it - if we fail to parse it, I'd just continue.

ryanpetrello commented:
> I see no reason we couldn't move ahead with this. @ryanpetrello I don't think you're going to like this, but I think we gotta do it.

I don't _love_ it, but I also don't think we really have a choice.

wenottingham commented:
Note that the only testing I've done so far is ensuring a change in the global ansible.cfg was reflected in the job env. This needs more test coverage.

AlanCoding commented on awx/main/utils/ansible.py:
```diff
@@ -97,3 +98,14 @@ def could_be_inventory(project_path, dir_path, filename):
+
+
+def read_ansible_config(project_path, variables_of_interest):
+    fnames = [ os.path.join(project_path, 'ansible.cfg'), '/etc/ansible/ansible.cfg']
```
unit tests are failing on the case where `project_path` is none.

https://github.com/ansible/awx/blob/b5fa1606bd14bc96d6f884dfee15bd38d06d8441/awx/main/models/projects.py#L190-L195

This would be an uncommon situation, but one that can happen in practice. The local_path is filled out in some logic around how the project is saved, and I suspect might not be there if it never completed an update, or some weird variation of that.

Anyway, if given a null project_path, then just use

```python
fnames = ['/etc/ansible/ansible.cfg']
```

Let something else in the task logic error instead of this.

AlanCoding commented:
My functional comments right now are:

 - thing about order of processing of the config files
 - handling null project path
 - (speculative) put whole thing in try-except

The latter point is where I have issues about what the expectation is. If there is a config which cannot be read, then Ansible core will error. If _we_ error in this logic, then it will be presented in a much less-clear way than:

```
$ ansible-playbook -i localhost, ~/Documents/repos/jlaska-ansible-playbooks/debug.yml 
Error reading config file (/Users/alancoding/Documents/repos/ansible/ansible.cfg): Source contains parsing errors: '<string>'
	[line  4]: '[foo\n'
```

IMHO. So I would rather this just not error. Doesn't matter what we do (other than logging it), just don't apply the config settings and move on.

AlanCoding commented:
Here is one test case that I believe is perfectly valid, but maybe doesn't test this issue directly.

https://github.com/ansible/test-playbooks/pull/79

will get a case with exactly what was mentioned in the issue.

wenottingham commented:
re: suggestions
- Order appears correct. Ansible is first found, with an order of env var specified (which we're ignoring for this), CWD, homedir (which we're ignoring), and then system.
- null project path: fixed
- try/except: added a check that we actually parsed a 'defaults' section. ConfigParser appears to eat all errors if passed nonexistent files, or non-config files, or unreadable files, so we're good there

wenottingham committed:
Wrap ansible.cfg reading so that any errors aren't fatal.

AlanCoding reviewed (approved):
I guess my comments are minor enough that we can get by without fixing it now. We just need testing from here.

AlanCoding commented:
Snag encountered while testing this: https://github.com/ansible/ansible/issues/63341

AlanCoding commented:
Link https://github.com/ansible/test-playbooks/pull/84

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -1529,14 +1530,22 @@ def build_env(self, job, private_data_dir, isolated=False, private_data_files=No
-                ('ANSIBLE_COLLECTIONS_PATHS', 'requirements_collections', '~/.ansible/collections:/usr/share/ansible/collections'),
-                ('ANSIBLE_ROLES_PATH', 'requirements_roles', '~/.ansible/roles:/usr/share/ansible/roles:/etc/ansible/roles')):
+        path_vars = (
+            ('ANSIBLE_COLLECTIONS_PATHS', 'collections_path', 'requirements_collections', '~/.ansible/collections:/usr/share/ansible/collections'),
```
https://github.com/ansible/ansible/blob/devel/lib/ansible/config/base.yml#L226

`collections_path` -> `collections_paths`

AlanCoding reviewed (changes requested):
got setting name wrong, otherwise I believe this should be closing in on functionality
